ccmbioinfo/lfs
# CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

**Please note that this PR is based on `CARDS-1892` (currently open) and thus it will remain in _draft_ mode until `CARDS-1892` is merged into `dev`. Furthermore, testing this PR requires the tests from `CARDS-1892` to have been completed.**

This PR implements CARDS-1896 by providing the `scan_remote_package_list.sh` script which (along with its dependencies introduced in this PR) queries a GitHub repository, via the GitHub App API, for the set of OS packages installed on a virtual machine. The script then builds a very simple Docker image (as a TAR file) containing the list of OS packages obtained from the GitHub App API and a few other small supporting files. The Docker image TAR file is then analyzed using the `aquasec/trivy` security scanner and the scan results are parsed and formatted into a Slack message.

### Setup / Testing Instructions

1. Create a new GitHub App by going to https://github.com/settings/apps and click the _New GitHub App_ button.
2. Give your application some random name, set the _Homepage URL_ your GitHub page (eg. https://github.com/IntegralProgrammer), disable _Expire user authentication tokens_, deactivate the _webhook_, set the _Contents_ permission to `Read-only`

![contents_read_only_permission](https://user-images.githubusercontent.com/5349728/186735646-d43bf336-6947-43a8-9fa2-423b908e4ade.png)

Set _Where can this GitHub App be installed?_ to _Only on this account_ and click _Create GitHub App_.

3. Click the _Generate a private key_ button and save the private key `.pem` file.
4. Make note of the _App ID_.
5. Install the App into the `CARDS-1892-test` repository (the repository that was created when testing CARDS-1892).
6. Make note of the _installation ID_ from the URL in the address bar `https://github.com/settings/installations/<INSTALLATION ID IS HERE>`.
7. Configure the following environment variables
```bash
export GITHUB_API_PRIVATE_KEY=$(cat /path/to/private/key.pem/from/step/3)
export GITHUB_API_APP_ID="<App ID from step 4>"
export GITHUB_API_INSTALLATION_ID="<Installation ID from step 6>"
export GITHUB_REPOSITORY="<Your-GitHub-Username/CARDS-1892-test>"
```
8. Enter the `Utilities/Administration/trivy-utils` directory and run the command `./scan_remote_package_list.sh cardsdemo ~/example_slack_report.json`.
9. Once the command completes, verify that the `~/example_slack_report.json` file has been populated.
10. Enter the `Utilities/Administration` directory and run the command `python3 generate_slack_vulnerability_report_for_host.py --deployment_hostname cardsdemo --vm_os_json_file ~/example_slack_report.json --output ~/example_slack_blocks.json`.
11. Send the contents of `~/example_slack_blocks.json` to a development/testing channel by setting the `SLACK_INCOMING_WEBHOOK_URL` environment variable to the incoming webhook URL for your testing channel or, _alternatively_, send me a Slack message and I can add you to the `cards-alerts-dev` channel and give you the value for `SLACK_INCOMING_WEBHOOK_URL`. Once this environment variable has been configured, in the `Utilities/Administration` directory run `python3 post_message_blocks_to_slack.py ~/example_slack_blocks.json`.
12. A message resembling the following should now be shown on the Slack channel.
![Screenshot_2022-08-25_15-06-28](https://user-images.githubusercontent.com/5349728/186748364-74a27221-68f2-4448-91e4-1492eb9c509b.png)
